### ISSUE TITLE
Docker for Windows Container are not reachable outside host anymore

### ISSUE THREAD

**Issue by zhujik** [docker/for-win#4391]
- [x] I have tried with the latest version of my channel (Stable or Edge)
  - [x] I have uploaded Diagnostics
  - Diagnostics ID:
58CFA20E-7891-4FB1-B79E-17F960480329/20190802141658

### Expected behavior
Docker Containers are from other machines reachable via computername:port from the network

### Actual behavior
Docker containers are not reachable from other machines anymore

### Information
  - Windows Version: 10
  - Docker Desktop Version: 2.1

### Steps to reproduce the behavior

  1. Run a container on a specific port, e.g. docker run --name mynginx1 -p 8080:80 -d nginx
  2. From another machine in the network. type the following in the browser: http://computername:8080/
  3. Expect to see nginx welcome page, instead timeout. Same happens for IP
  4. From the host itself the container is reachable as normal (via localhost and computername and ip)
  5. Issue is there since version 2.1 upgrade, it works on machines that still run the old docker desktop version

**Comment by mikeparker** [comment]
I can't see anything special from your logs other than the fact you have multiple DNS servers. Can you check your DNS settings haven't been reset and look correct from the settings page?

Also I assume you haven't switched between linux and windows container mode?

**Comment by zhujik** [comment]
Hi mikeparker, thank you for your answer.

The DNS settings look normal to me. DNS server is set to automatic and has ever been. I didn't switch between container modes, it's linux for all hosts that we tested this on. So far for all computers that we updated in our department (4), this issue materialized.

Which information could I provide to you to help debug this issue?

**Comment by sjdvda** [comment]
I am having exactly the same issue (Windows 10, Docker 2.1 Edge and Stable). What I have tried so far:
* Reinstalling Docker
* Disabling Hyper-V and enabling it again
* Resetting all network interfaces



Nothing has worked so far. I can only access my containers from the local machine. Other (non-Docker) programs can be accessed from the LAN so it's definitely a Docker issue.

**Comment by erickhchan** [comment]
Having the same issue with Windows 10, Version 2.1.0.0 (36874) after updating just a few hours ago. 

My docker containers aren't able to resolve internal hostname. When using the resolved ip address everything works as expected.

**Comment by kjlhgfds** [comment]
I downgraded windows docker and the issue is gone, so definitly an upgrade issue.

**Comment by sjdvda** [comment]
> 
> I downgraded windows docker and the issue is gone, so definitly an upgrade related issue.

Did you have to uninstall docker to be able to downgrade? And did you lose your containers when uninstalling?

**Comment by frankyifei** [comment]
I have the same problem. Is there a way to manually set the internal dns?

**Comment by kjlhgfds** [comment]
> Did you have to uninstall docker to be able to downgrade? And did you lose your containers when uninstalling?

Yes I did because I have less than 10 containers.
If someone has a way to save volumes outside dockers so we can bakcup them and put them again in the new docker installation, I'd be glade to hear you !

**Comment by djs55** [comment]
@zhujik In the diagnostics I see the container starting and opening a port forward as expected
```
[16:08:02.627][GoBackendProcess  ][Info   ] Adding tcp forward from 0.0.0.0:80 to 172.17.0.2:80
```
and the port forwarder has the following:
```json
{"proto":"tcp","out_ip":"0.0.0.0","out_port":80,"out_path":"","in_ip":"172.17.0.2","in_port":80,"in_path":"","annotation":""}
```
Given that it works on the local machine but not on remote machines I think it must be some kind of firewall issue. Could you try disabling your firewall and trying again?

I attempted to reproduce the problem locally but it works as expected for me (unfortunately).

**Comment by sjdvda** [comment]
I have disabled my firewall but unfortunately the issue still persists.

Diagnostic ID: 3AF31CA1-733A-4D1C-972E-46213C91A1AA/20190806175545

**Comment by frankyifei** [comment]
same here
AD945C75-43E0-4F00-8D23-FE2061E742D0/20190807082423

**Comment by bigtongue5566** [comment]
same problem.
I disabled windows firewall for public network and it works for me.

**Comment by zhujik** [comment]
@djs55 thank you for your answer.
When the domain firewall is disabled, the issue is resolved. However, since turning off the domain firewall imposes a security issue, this is not a permanent workaround. What do I need to configure in this firewall to resolve this issue?

Still, I don't know why the domain firewall blocks containers that run with docker desktop 2.1 and not for older versions?!

**Comment by sjdvda** [comment]
Disabling Windows Firewall for both Private networks and Public networks made containers reachable again for me as well. But as @zhujik said, this is not a permanent workaround.

**Comment by sjdvda** [comment]
I noticed that the program name changed from "Docker for Windows" to "Docker Desktop" between 2.0 and 2.1 and so did the executable. I checked the Advanced Windows Firewall rules and I noticed that there was an entry to allow all inbound and outbound connections for "C:\Program Files\Docker\Docker\Docker for Windows.exe" (a file that no longer exists) but no such rule for "C:\Program Files\Docker\Docker\Docker Desktop.exe". 

Unfortunately, creating new inbound and outbound "Allow all" rules in Windows Firewall for "Docker Desktop.exe" does not make a difference for me. My containers are still only reachable from the host machine.

**Comment by sjdvda** [comment]
After a lot of trial and error, I believe I have solved the issue:

In Windows Defender Firewall with Advanced Security, the following rule needs to be created:

> Type: Inbound
> Program: C:\Program Files\Docker\Docker\resources\com.docker.backend.exe
> Allow all connections

Other machines on the local network can now reach my containers. My nginx reverse proxy is also working again and I can access my exposed containers from the internet. Hopefully this fix will also work for others in this thread.

**Comment by simonfagerholm** [comment]
Had the same issue, and solved it the same way as @sjdvda (unfortunately I didn't find this issue until after).
I used `netstat -nab` to figure out which application was using the ports that where configured for the docker containers. In the issue above it would be 8080. In case this happens again, the same steps should usable to resolse the issue.

**Comment by kjlhgfds** [comment]
The solution given by @sjdvda worked for me, then suddenly, without changing/upgrading anything, it doesn't work anymore.
Can anyone help me ?

**Comment by kkbruce** [comment]
Diagnostics ID: **DFB31BFA-698D-4A0A-8595-C83F3C941F9B/20191104055931**

Like title problem. My Windows Container can't find DNS record.

I have check [issuecomment-520420942](https://github.com/docker/for-win/issues/4391#issuecomment-520420942 ) `com.docker.backend.exe` for us anti-virus SEP.

## Expected behavior

Get Domain A record from DNS server.

## Actual behavior

Can't get Domain A record from DNS server.

## Information

Windows version:

```
ver

Microsoft Windows [版本 10.0.18362.418]
```

Docker version:

```
docker version
Client: Docker Engine - Community
 Version:           19.03.4
 API version:       1.40
 Go version:        go1.12.10
 Git commit:        9013bf5
 Built:             Thu Oct 17 23:44:48 2019
 OS/Arch:           windows/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.4
  API version:      1.40 (minimum version 1.24)
  Go version:       go1.12.10
  Git commit:       9013bf5
  Built:            Thu Oct 17 23:55:51 2019
  OS/Arch:          windows/amd64
  Experimental:     false
```

## Steps to reproduce the behavior

Run `docker run -it --rm mcr.microsoft.com/windows/servercore:1903 cmd`

```
C:\>ping www.google.com

Pinging www.google.com [172.217.160.68] with 32 bytes of data:
Control-C
^C
C:\>nslookup www.google.com
Server:  UnKnown
Address:  172.31.16.1

*** UnKnown can't find www.google.com: Server failed
C:\>powershell
PS C:\> Invoke-WebRequest -Uri http://www.google.com -UseBasicParsing
Invoke-WebRequest : Unable to connect to the remote server
At line:1 char:1
+ Invoke-WebRequest -Uri http://www.google.com -UseBasicParsing
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-WebRequest], WebException
    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand
PS C:\> Invoke-WebRequest -Uri https://www.google.com -UseBasicParsing
Invoke-WebRequest : Unable to connect to the remote server
At line:1 char:1
+ Invoke-WebRequest -Uri https://www.google.com -UseBasicParsing
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-WebRequest], WebException
    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand
```

If I run `docker run` command, I found many `The user name or password is incorrect` error message from EventLog:

```
hcsshim::ComputeSystem::CreateProcess - End Operation - Error [cid=d212c890f86a5162214de2bee3c2db6d19d5ca61f0041013bad9380605798677 error=CreateProcess d212c890f86a5162214de2bee3c2db6d19d5ca61f0041013bad9380605798677: The user name or password is incorrect.
[Event Detail:  Provider: 00000000-0000-0000-0000-000000000000]
[Event Detail:  Provider: 00000000-0000-0000-0000-000000000000]
[Event Detail: onecore\vm\compute\management\orchestration\vmhostedcontainer\processmanagement.cpp(173)\vmcomputeagent.exe!00007FF6B2E9AE77: (caller: 00007FF6B2E4E4AB) Exception(2) tid(380) 8007052E The user name or password is incorrect.
    CallContext:[\Bridge_ProcessMessage\VmHostedContainer_ExecuteProcess] 
 Provider: 00000000-0000-0000-0000-000000000000]
(extra info: {"CommandLine":"cmd.exe /C \"ECHO 172.21.35.173    host.docker.internal \u003e\u003e %systemroot%\\system32\\drivers\\etc\\hosts \u0026 ECHO 172.21.35.173    gateway.docker.internal \u003e\u003e %systemroot%\\system32\\drivers\\etc\\hosts\"","User":"Administrator","WorkingDirectory":"/","CreateStdInPipe":true,"CreateStdOutPipe":true,"CreateStdErrPipe":true,"ConsoleSize":[0,0]})]
```
---
```
exec's CreateProcess() failed [namespace=moby exec=aeedcfbc0ce7b9404c576a1524ca69e70f40c15d3b0e6b118862b1115e6410b1 container=d212c890f86a5162214de2bee3c2db6d19d5ca61f0041013bad9380605798677 module=libcontainerd error=container d212c890f86a5162214de2bee3c2db6d19d5ca61f0041013bad9380605798677 encountered an error during CreateProcess: failure in a Windows system call: The user name or password is incorrect. (0x52e)
[Event Detail:  Provider: 00000000-0000-0000-0000-000000000000]
[Event Detail:  Provider: 00000000-0000-0000-0000-000000000000]
[Event Detail: onecore\vm\compute\management\orchestration\vmhostedcontainer\processmanagement.cpp(173)\vmcomputeagent.exe!00007FF6B2E9AE77: (caller: 00007FF6B2E4E4AB) Exception(2) tid(380) 8007052E The user name or password is incorrect.
    CallContext:[\Bridge_ProcessMessage\VmHostedContainer_ExecuteProcess] 
 Provider: 00000000-0000-0000-0000-000000000000] extra info: {"CommandLine":"cmd.exe /C \"ECHO 172.21.35.173    host.docker.internal \u003e\u003e %systemroot%\\system32\\drivers\\etc\\hosts \u0026 ECHO 172.21.35.173    gateway.docker.internal \u003e\u003e %systemroot%\\system32\\drivers\\etc\\hosts\"","User":"Administrator","WorkingDirectory":"/","CreateStdInPipe":true,"CreateStdOutPipe":true,"CreateStdErrPipe":true,"ConsoleSize":[0,0]}]
```
---
```
Error running exec aeedcfbc0ce7b9404c576a1524ca69e70f40c15d3b0e6b118862b1115e6410b1 in container: container d212c890f86a5162214de2bee3c2db6d19d5ca61f0041013bad9380605798677 encountered an error during CreateProcess: failure in a Windows system call: The user name or password is incorrect. (0x52e)
[Event Detail:  Provider: 00000000-0000-0000-0000-000000000000]
[Event Detail:  Provider: 00000000-0000-0000-0000-000000000000]
[Event Detail: onecore\vm\compute\management\orchestration\vmhostedcontainer\processmanagement.cpp(173)\vmcomputeagent.exe!00007FF6B2E9AE77: (caller: 00007FF6B2E4E4AB) Exception(2) tid(380) 8007052E The user name or password is incorrect.
    CallContext:[\Bridge_ProcessMessage\VmHostedContainer_ExecuteProcess] 
 Provider: 00000000-0000-0000-0000-000000000000] extra info: {"CommandLine":"cmd.exe /C \"ECHO 172.21.35.173    host.docker.internal \u003e\u003e %systemroot%\\system32\\drivers\\etc\\hosts \u0026 ECHO 172.21.35.173    gateway.docker.internal \u003e\u003e %systemroot%\\system32\\drivers\\etc\\hosts\"","User":"Administrator","WorkingDirectory":"/","CreateStdInPipe":true,"CreateStdOutPipe":true,"CreateStdErrPipe":true,"ConsoleSize":[0,0]}
```
---
```
hcsshim::Process::CloseStdin - End Operation - Error [cid=d212c890f86a5162214de2bee3c2db6d19d5ca61f0041013bad9380605798677 pid=1244 error=hcsshim::Process::CloseStdin d212c890f86a5162214de2bee3c2db6d19d5ca61f0041013bad9380605798677:1244: hcsshim: the handle has already been closed]
```

**Comment by DougSchmidt-AI** [comment]
Just had the same problem.

Was running Docker for Windows 2.1.0.5, and my containers could be accessed over the network.
I upgraded to v2.2.0.0 and none of my containers could be accessed over the network, only via localhost.

I tried the @sjdvda workaround, but no luck. After an hour of fidlling, my only option was to uninstall v2.2.0.0 and reinstall v2.1.0.5. My containers were now accessible again.

**Comment by gmerciel** [comment]
@DougSchmidt-AI I had the same problem. Search for a blocking rule call com.docker.bakend added after upgrade. Blocking takes precedence, that's why it doesn't work even if you the rule in @sjdvda workaround.

**Comment by Neurrone** [comment]
@gmerciel thanks so much, I wasted hours trying to figure out why services in Docker containers wouldn't be reachable outside localhost even though I've already added a rule to open the right ports.

The question then is whether these rules are automatically added by default somehow - if so, this would be a pretty annoying problem. 

Windows firewall itself is also broken because even with logging enabled, the dropped packets arent even logged, making troubleshooting extremely difficult.
